File: Android/app/src/main/java/com/agora/entfulldemo/home/HomeMineFragment.java
package com.agora.entfulldemo.home;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.FileProvider;
import androidx.lifecycle.ViewModelProvider;

import com.agora.entfulldemo.BuildConfig;
import com.agora.entfulldemo.R;
import com.agora.entfulldemo.databinding.AppFragmentHomeMineBinding;
import com.agora.entfulldemo.home.mine.AboutUsActivity;

import java.io.File;

import io.agora.rtc2.RtcEngine;
import io.agora.scene.base.Constant;
import io.agora.scene.base.GlideApp;
import io.agora.scene.base.api.model.User;
import io.agora.scene.base.component.AgoraApplication;
import io.agora.scene.base.component.BaseViewBindingFragment;
import io.agora.scene.base.component.OnButtonClickListener;
import io.agora.scene.base.manager.PagePilotManager;
import io.agora.scene.base.manager.UserManager;
import io.agora.scene.base.utils.FileUtils;
import io.agora.scene.base.utils.SPUtil;
import io.agora.scene.base.utils.ToastUtils;
import io.agora.scene.base.utils.UriUtils;
import io.agora.scene.widget.dialog.CommonDialog;
import io.agora.scene.widget.dialog.EditNameDialog;
import io.agora.scene.widget.dialog.SelectPhotoFromDialog;
import io.agora.scene.widget.utils.CenterCropRoundCornerTransform;
import io.agora.scene.widget.utils.ImageCompressUtil;

public class HomeMineFragment extends BaseViewBindingFragment<AppFragmentHomeMineBinding> {
    private CommonDialog logoutDialog;
    private CommonDialog logoffAccountDialog;
    private MainViewModel mainViewModel;
    private SelectPhotoFromDialog selectPhotoFromDialog;
    private EditNameDialog editNameDialog;

    private CommonDialog debugModeDialog;
    private int counts = 0;
    private final long debugModeOpenTime = 2000;
    private long beginTime = 0;

    @NonNull
    @Override
    protected AppFragmentHomeMineBinding getViewBinding(@NonNull LayoutInflater inflater, @Nullable ViewGroup container) {
        return AppFragmentHomeMineBinding.inflate(inflater);
    }


    @Override
    public void initView() {
        String versionString = "20230530-" + io.agora.scene.base.BuildConfig.APP_VERSION_NAME + "-" + RtcEngine.getSdkVersion();
        mainViewModel = new ViewModelProvider(this).get(MainViewModel.class);
        mainViewModel.setLifecycleOwner(this);
        getBinding().tvVersion.setText(getString(R.string.app_mine_current_version, versionString));
    }

    @SuppressLint("SetTextI18n")
    @Override
    public void initListener() {
        mainViewModel.setISingleCallback((type, o) -> {
            if (type == Constant.CALLBACK_TYPE_USER_INFO_CHANGE) {
                User user = UserManager.getInstance().getUser();
                GlideApp.with(this).load(user.headUrl).error(R.mipmap.userimage)
                        .transform(new CenterCropRoundCornerTransform(100))
                        .into(getBinding().ivUserAvatar);
                getBinding().tvUserID.setText(getString(R.string.id_is_, user.userNo));
                getBinding().tvUserMobile.setText(user.name);
            } else if (type == Constant.CALLBACK_TYPE_USER_CANCEL_ACCOUNTS) {
                UserManager.getInstance().logout();
                requireActivity().finish();
                PagePilotManager.pageWelcome();
            }
        });
        getBinding().tvUserAgreement.setOnClickListener(view -> {
            PagePilotManager.pageWebView("https://accktvpic.oss-cn-beijing.aliyuncs.com/pic/meta/demo/fulldemoStatic/privacy/service.html");
        });
        getBinding().tvPrivacyAgreement.setOnClickListener(view -> {
            PagePilotManager.pageWebView("https://accktvpic.oss-cn-beijing.aliyuncs.com/pic/meta/demo/fulldemoStatic/privacy/privacy.html");
        });

        getBinding().tvLogout.setOnClickListener(view -> {
            showLogoutDialog();
        });
        getBinding().tvLogoffAccount.setOnClickListener(view -> {
            showLogoffAccountDialog();
        });
        getBinding().tvAbout.setOnClickListener(view -> {
            startActivity(new Intent(getContext(), AboutUsActivity.class));
        });
        getBinding().vToEdit.setOnClickListener(view -> {
            if (editNameDialog == null) {
                editNameDialog = new EditNameDialog(getContext());
                editNameDialog.iSingleCallback = (type, o) -> {
                    if (type == 0) {
                        mainViewModel.requestEditUserInfo(null, (String) o, null);
                    }
                };
            }
            editNameDialog.show();
        });
        getBinding().ivUserAvatar.setOnClickListener(view -> {
            ((MainActivity) requireActivity()).requestReadStoragePermission(true);
        });
        getBinding().tvVersion.setOnClickListener(v -> {
            if (counts == 0) {
                beginTime = System.currentTimeMillis();
            }
            counts++;
            if (counts == 5) {
                if (System.currentTimeMillis() - beginTime > debugModeOpenTime) {
                    counts = 0;
                    return;
                }
                counts = 0;
                getBinding().tvDebugMode.setVisibility(View.VISIBLE);
                AgoraApplication.the().enableDebugMode(true);
                ToastUtils.showToast("Debug模式已打开");
            }
        });
        getBinding().tvDebugMode.setOnClickListener(v -> showDebugModeCloseDialog());
        if (AgoraApplication.the().isDebugModeOpen()) {
            getBinding().tvDebugMode.setVisibility(View.VISIBLE);
        }
    }


    private static final int CHOOSE_PHOTO = 100;
    private static final int TAKE_PHOTO = 101;
    String mTempPhotoPath = null;

    private void showSelectPhotoFromDialog() {
        if (selectPhotoFromDialog == null) {
            selectPhotoFromDialog = new SelectPhotoFromDialog(getContext());
            selectPhotoFromDialog.setOnButtonClickListener(new OnButtonClickListener() {
                @Override
                public void onLeftButtonClick() {
                    openAlbum();
                }

                @Override
                public void onRightButtonClick() {
                    takePhoto();
                }
            });
        }
        selectPhotoFromDialog.show();
    }

    public void openAlbum() {
        Intent intentToPickPic = new Intent(Intent.ACTION_PICK, null);
        intentToPickPic.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        intentToPickPic.setDataAndType(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, "image/*");
        startActivityForResult(intentToPickPic, CHOOSE_PHOTO);
    }

    private void takePhoto() {
        Intent intentToTakePhoto = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
        intentToTakePhoto.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
        File fileDir = new File(FileUtils.getTempSDPath()
        );
        if (!fileDir.exists()) {
            fileDir.mkdirs();
        }
        File photoFile = new File(fileDir, "photo.jpg");
        mTempPhotoPath = photoFile.getAbsolutePath();
        Uri imageUri = FileProvider.getUriForFile(
                getContext(),
                BuildConfig.APPLICATION_ID + ".fileProvider",
                photoFile
        );
        intentToTakePhoto.putExtra(MediaStore.EXTRA_OUTPUT, imageUri);
        startActivityForResult(intentToTakePhoto, TAKE_PHOTO);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == CHOOSE_PHOTO) {
                Uri uri = data.getData();
                if (uri != null) {
                    String filePath = UriUtils.INSTANCE.getFilePathByUri(getContext(), uri);
                    if (!TextUtils.isEmpty(filePath)) {
                        setImage(filePath);
                    }
                }
            } else if (requestCode == TAKE_PHOTO) {
                setImage(mTempPhotoPath);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private void setImage(String filePath) {
        if (filePath == null) return;
        String path = ImageCompressUtil.displayPath(requireActivity(), filePath);
        if (TextUtils.isEmpty(path) || new File(filePath).length() <= 150000) {
            path = filePath;
        }
        mTempPhotoPath = path;
//        UserManager.getInstance().getUser().headUrl = mTempPhotoPath;
        mainViewModel.updatePhoto(new File(mTempPhotoPath));
    }

    @Override
    public void requestData() {
        mainViewModel.requestUserInfo(UserManager.getInstance().getUser().userNo);
    }

    private void showLogoffAccountDialog() {
        if (logoffAccountDialog == null) {
            logoffAccountDialog = new CommonDialog(requireContext());
            logoffAccountDialog.setDialogTitle("确定注销账号？");
            logoffAccountDialog.setDescText("注销账号后，您将暂时无法使用该账号体验我们的服务，真的要注销吗？");
            logoffAccountDialog.setDialogBtnText(getString(R.string.app_logoff), getString(R.string.cancel));
            logoffAccountDialog.setOnButtonClickListener(new OnButtonClickListener() {
                @Override
                public void onLeftButtonClick() {
                    SPUtil.putBoolean(Constant.IS_AGREE, false);
                    mainViewModel.requestCancellation(UserManager.getInstance().getUser().userNo);
                }

                @Override
                public void onRightButtonClick() {
                    SPUtil.putBoolean(Constant.IS_AGREE, true);
                }
            });
        }
        logoffAccountDialog.show();
    }

    private void showLogoutDialog() {
        if (logoutDialog == null) {
            logoutDialog = new CommonDialog(requireContext());
            logoutDialog.setDialogTitle("确定退出登录吗？");
            logoutDialog.setDescText("退出登录后，我们还会继续保留您的账户数据，记得再来体验哦～");
            logoutDialog.setDialogBtnText(getString(R.string.app_exit), getString(R.string.cancel));
            logoutDialog.setOnButtonClickListener(new OnButtonClickListener() {
                @Override
                public void onLeftButtonClick() {
                    SPUtil.putBoolean(Constant.IS_AGREE, false);
                    UserManager.getInstance().logout();
                    requireActivity().finish();
                    PagePilotManager.pageWelcome();
                }

                @Override
                public void onRightButtonClick() {

                }
            });
        }
        logoutDialog.show();
    }

    private void showDebugModeCloseDialog() {
        if (debugModeDialog == null) {
            debugModeDialog = new CommonDialog(requireContext());
            debugModeDialog.setDialogTitle("确定退出Debug模式么？");
            debugModeDialog.setDescText("退出debug模式后， 设置页面将恢复成正常的设置页面哦～");
            debugModeDialog.setDialogBtnText(getString(R.string.cancel), getString(R.string.app_exit));
            debugModeDialog.setOnButtonClickListener(new OnButtonClickListener() {
                @Override
                public void onLeftButtonClick() {
                }

                @Override
                public void onRightButtonClick() {
                    counts = 0;
                    getBinding().tvDebugMode.setVisibility(View.GONE);
                    AgoraApplication.the().enableDebugMode(false);
                    ToastUtils.showToast("Debug模式已关闭");
                }
            });
        }
        debugModeDialog.show();
    }
}

File: Android/app/src/main/java/com/agora/entfulldemo/home/HomeMineFragment.kt
package com.agora.entfulldemo.home

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.graphics.Rect
import android.os.Bundle
import android.provider.MediaStore
import android.text.InputFilter
import android.text.TextUtils
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.FileProvider
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isVisible
import androidx.lifecycle.ViewModelProvider
import com.agora.entfulldemo.databinding.AppFragmentHomeMineBinding
import com.agora.entfulldemo.home.constructor.URLStatics
import com.agora.entfulldemo.home.mine.AppDebugActivity
import com.bumptech.glide.request.RequestOptions
import io.agora.scene.base.BuildConfig
import io.agora.scene.base.Constant
import io.agora.scene.base.GlideApp
import io.agora.scene.base.component.AgoraApplication
import io.agora.scene.base.component.BaseViewBindingFragment
import io.agora.scene.base.component.OnButtonClickListener
import io.agora.scene.base.component.OnFastClickListener
import io.agora.scene.base.manager.PagePilotManager
import io.agora.scene.base.manager.UserManager
import io.agora.scene.base.utils.FileUtils
import io.agora.scene.base.utils.UriUtils
import io.agora.scene.widget.dialog.SelectPhotoFromDialog
import io.agora.scene.widget.utils.ImageCompressUtil
import java.io.File

class HomeMineFragment : BaseViewBindingFragment<AppFragmentHomeMineBinding>() {

    companion object {
        private const val CHOOSE_PHOTO = 100
        private const val TAKE_PHOTO = 101
    }

    private val mainViewModel: MainViewModel by lazy {
        ViewModelProvider(this)[MainViewModel::class.java]
    }

    private var selectPhotoFromDialog: SelectPhotoFromDialog? = null
    override fun getViewBinding(inflater: LayoutInflater, container: ViewGroup?): AppFragmentHomeMineBinding {
        return AppFragmentHomeMineBinding.inflate(inflater)
    }

    //防止多次回调
    private var lastKeyBoard = false

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ViewCompat.setOnApplyWindowInsetsListener(view) { v: View?, insets: WindowInsetsCompat ->
            val inset = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            view.setPaddingRelative(inset.left, inset.top, inset.right, 0)
            WindowInsetsCompat.CONSUMED
        }
        activity?.window?.let { window ->
            // 获取根布局可见区域的高度
            val initialWindowHeight = Rect().apply { window.decorView.getWindowVisibleDisplayFrame(this) }.height()
            view.viewTreeObserver.addOnGlobalLayoutListener {
                val tempWindow = activity?.window ?: return@addOnGlobalLayoutListener
                val currentWindowHeight =
                    Rect().apply { tempWindow.decorView.getWindowVisibleDisplayFrame(this) }.height()
                // 判断键盘高度来确定键盘的显示状态
                if (currentWindowHeight < initialWindowHeight) {
                    if (lastKeyBoard) return@addOnGlobalLayoutListener
                    lastKeyBoard = true
                    binding.etNickname.selectAll()
                    // 软键盘可见
                    Log.d("zhangw", "current: $currentWindowHeight, initial: $initialWindowHeight, show: true")
                    binding.ivEditNickname.isVisible = false

                } else {
                    if (!lastKeyBoard) return@addOnGlobalLayoutListener
                    lastKeyBoard = false
                    // 软键盘已收起
                    Log.d("zhangw", "current: $currentWindowHeight, initial: $initialWindowHeight, show: false")
                    binding.ivEditNickname.isVisible = true
                    binding.etNickname.clearComposingText()
                    binding.etNickname.clearFocus()
                    val newName = binding.etNickname.text.toString()
                    if (newName.isEmpty()) {
                        binding.etNickname.setText(UserManager.getInstance().user.name)
                    } else if (newName != UserManager.getInstance().user.name) {
                        mainViewModel.requestEditUserInfo(null, newName, null)
                    }
                }
            }
        }
    }

    override fun initView() {
        mainViewModel.setLifecycleOwner(this)
        mainViewModel.setISingleCallback { type, data ->
            if (type == Constant.CALLBACK_TYPE_USER_INFO_CHANGE || type == Constant.CALLBACK_TYPE_REQUEST_USER_INFO) {
                val user = UserManager.getInstance().user
                GlideApp.with(this)
                    .load(user.headUrl)
                    .placeholder(io.agora.scene.widget.R.mipmap.default_user_avatar)
                    .error(io.agora.scene.widget.R.mipmap.default_user_avatar)
                    .apply(RequestOptions.circleCropTransform())
                    .into(binding.ivUserAvatar)
                binding.tvUserPhone.text = hidePhoneNumber(user.mobile)
                binding.etNickname.setText(user.name)
            }
        }
    }

    override fun onResume() {
        super.onResume()
        binding.tvDebugMode.isVisible = AgoraApplication.the().isDebugModeOpen
    }

    @SuppressLint("SetTextI18n")
    override fun initListener() {
        binding.tvMineAccount.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                PagePilotManager.pageMineAccount()
            }
        })
        binding.tvUserAgreement.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                PagePilotManager.pageWebView(URLStatics.userAgreementURL)
            }
        })
        binding.tvPrivacyAgreement.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                PagePilotManager.pageWebView(URLStatics.privacyAgreementURL)
            }
        })
        binding.tvCollectionChecklist.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                val stringBuilder =
                    StringBuilder(URLStatics.collectionChecklistURL)
                        .append("?userNo=").append(UserManager.getInstance().user.userNo)
                        .append("&appId=").append(BuildConfig.AGORA_APP_ID)
                        .append("&projectId=")
                        .append("agora_ent_demo") //.append("&sceneId=").append("-1")
                        .append("&token=").append(UserManager.getInstance().user.token)
                PagePilotManager.pageWebView(stringBuilder.toString())
            }
        })
        binding.tvThirdDataSharing.setOnClickListener(object : OnFastClickListener() {

            override fun onClickJacking(view: View) {
                PagePilotManager.pageWebView(URLStatics.thirdDataSharingURL)
            }
        })
        binding.tvAbout.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                PagePilotManager.pageAboutUs()
            }
        })
        binding.tvFeedback.setOnClickListener(object : OnFastClickListener() {
            override fun onClickJacking(view: View) {
                PagePilotManager.pageFeedback()
            }
        })
        binding.etNickname.setOnFocusChangeListener { v, hasFocus ->
            Log.d("zhangw", "etNickname setOnFocusChangeListener hasFocus $hasFocus")
        }
        binding.ivEditNickname.setOnClickListener {
            binding.etNickname.requestFocus()
            showKeyboard(binding.etNickname)
        }
        binding.ivUserAvatar.setOnClickListener { view: View? ->
            (requireActivity() as MainActivity).requestReadStoragePermission(
                true
            )
        }
        binding.tvDebugMode.setOnClickListener { v: View? ->
            activity?.let { activity->
                AppDebugActivity.startActivity(activity)
            }
        }
        binding.tvDebugMode.isVisible = AgoraApplication.the().isDebugModeOpen
        binding.etNickname.filters = arrayOf(mInputFilter)
    }

    private var mTempPhotoPath: String? = null
    private fun showSelectPhotoFromDialog() {
        val cxt = context ?: return
        if (selectPhotoFromDialog == null) {

            selectPhotoFromDialog = SelectPhotoFromDialog(cxt)
            selectPhotoFromDialog?.onButtonClickListener = object : OnButtonClickListener {
                override fun onLeftButtonClick() {
                    openAlbum()
                }

                override fun onRightButtonClick() {
                    takePhoto()
                }
            }
        }
        selectPhotoFromDialog?.show()
    }

    private fun hidePhoneNumber(phoneNumber: String): String {
        val prefix = phoneNumber.substring(0, 3)
        val suffix = phoneNumber.substring(7)
        return "$prefix****$suffix"
    }

    fun openAlbum() {
        val intentToPickPic = Intent(Intent.ACTION_PICK, null)
        intentToPickPic.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        intentToPickPic.setDataAndType(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, "image/*")
        startActivityForResult(intentToPickPic, CHOOSE_PHOTO)
    }

    private fun takePhoto() {
        val cxt = context ?: return
        val intentToTakePhoto = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
        intentToTakePhoto.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        val fileDir = File(
            FileUtils.getTempSDPath()
        )
        if (!fileDir.exists()) {
            fileDir.mkdirs()
        }
        val photoFile = File(fileDir, "photo.jpg")
        mTempPhotoPath = photoFile.absolutePath
        val imageUri = FileProvider.getUriForFile(
            cxt,
            com.agora.entfulldemo.BuildConfig.APPLICATION_ID + ".fileProvider",
            photoFile
        )
        intentToTakePhoto.putExtra(MediaStore.EXTRA_OUTPUT, imageUri)
        startActivityForResult(intentToTakePhoto, TAKE_PHOTO)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (resultCode == Activity.RESULT_OK) {
            if (requestCode == CHOOSE_PHOTO) {
                val uri = data?.data ?: return
                val cxt = context ?: return
                val filePath = UriUtils.INSTANCE.getFilePathByUri(cxt, uri)
                if (!TextUtils.isEmpty(filePath)) {
                    setImage(filePath)
                }
            } else if (requestCode == TAKE_PHOTO) {
                setImage(mTempPhotoPath)
            }
        }
        super.onActivityResult(requestCode, resultCode, data)
    }

    private fun setImage(filePath: String?) {
        if (filePath == null) return
        var path = ImageCompressUtil.displayPath(requireActivity(), filePath)
        if (TextUtils.isEmpty(path) || File(filePath).length() <= 150000) {
            path = filePath
        }
        mTempPhotoPath = path
        //        UserManager.getInstance().getUser().headUrl = mTempPhotoPath;
        mainViewModel.updatePhoto(File(mTempPhotoPath))
    }

    override fun requestData() {
        mainViewModel.requestUserInfo(UserManager.getInstance().user.userNo, true)
    }

    // 昵称限制 10 个字符
    private val maxLen = 10
    private val mInputFilter = InputFilter { source, start, end, dest, dstart, dend ->
        var dindex = 0
        var count = 0
        while (count <= maxLen && dindex < dest.length) {
            val c = dest[dindex++]
            count = if (c.code < 128) {
                count + 1
            } else {
                count + 2
            }
        }
        if (count > maxLen) {
            return@InputFilter dest.subSequence(0, dindex - 1)
        }
        var sindex = 0
        while (count <= maxLen && sindex < source.length) {
            val c = source[sindex++]
            count = if (c.code < 128) {
                count + 1
            } else {
                count + 2
            }
        }
        if (count > maxLen) {
            sindex--
        }
        source.subSequence(0, sindex)
    }
}